express_server.js

```diff
@@ -6,135 +6,175 @@ const cookieSession = require("cookie-session");
+    userID: "dummy",
+  },
 };
 
 //global variable, used to store and access users in the app
-let users = {  };
+const users = {};
+
+// Create an endpoint that renders an error info page.
+app.get("/errors/:statusCode", (req, res) => {
+  templateVars = {
+    statusCode: req.params.statusCode,
+    user: req.session.id,
+  };
+  res.render("url_errors", templateVars);
+});
 
 //registers the user (handles registration form data)
-app.post("/register", function(req, res) {
-  const newUserID = generateID();
+app.post("/register", (req, res) => {
   if (!req.body.email || !req.body.password) {
     // errors if fields are empty
-    res.status(400).send("Oops! Please enter an email and password :) <a href='/urls/'>Register</a>");
+    res.status(406).redirect("/errors/406");
   } else if (getUserByEmail(req.body.email, users)) {
-    res.send("Hey! We're already friends, just <a href='/login'>log in!</a> :)");
+    res.status(409).redirect("/errors/409");
   } else {
-    users[newUserID] = {id: newUserID, email: req.body.email, password: bcrypt.hashSync(req.body.password, 10)};
+    const newUserID = generateID();
+    users[newUserID] = {
+      id: newUserID,
+      email: req.body.email,
+      password: hashPassword(req.body.password),
+    };
     req.session.user_id = newUserID;
     res.redirect("/urls");
   }
 });
 
 //renders register template
-app.get('/register', function(req, res) {
-  let templateVars = { urls: urlDatabase, user:users[req.session.user_id] };
+app.get("/register", (req, res) => {
+  const templateVars = { user: req.session.user_id };
   res.render("user_registration", templateVars);
 });
 
-app.post('/login', function(req, res) {
-  if (!req.body.email || !req.body.password) {
-    // errors if fields are empty
-    res.status(400).send("Oops! Please enter an email and password :) <a href='/login'>Login</a>");
-  } else if (getUserByEmail(req.body.email, users) === undefined) {
-    // errors if user is not registered
-    res.status(403).send("Looks like we're not friends yet :( This email does not exist, please <a href='/register'>register!</a>");
-  } else if (getUserByEmail(req.body.email, users) && !bcrypt.compareSync(req.body.password, users[getUserByEmail(req.body.email, users)].password)) {
-    // errors if user is registered and passwords do not match
-    res.status(403).send("Oops! Password and email do not match! <a href='/login'>Login</a>");
-  } else if (getUserByEmail(req.body.email, users) && bcrypt.compareSync(req.body.password, users[getUserByEmail(req.body.email, users)].password)) {
+app.post("/login", (req, res) => {
+  const email = req.body.email;
+  const password = req.body.password;
+  if (
+    emailExists(users, email) &&
+    areValidEmailPassword(users, email, password)
+  ) {
+    // set the cookie
     req.session.user_id = getUserByEmail(req.body.email, users);
     res.redirect("/urls");
+  } else {
+    res.status(400).redirect("/errors/400");
   }
 });
```

Comment:
Simplified this one by refactoring the logic and using helpers.
`emailExists` and `areValidEmailPassword` (it is a good practice that functions that return a boolean should indicate this in their names, kinda like asking a yes or no question) are taking care of validating whatever was captured in those input fields so we can skip the check for empty.
If any of those validations fail then we just send an error.

See how by creating two helper functions the code is more declarative (you can see those two functions are validating email and password but you don't see how) and therefore easier to read. In the previous version there were a lot of ifs and repeated calls to bcrypt which makes it harder to follow. That logic is considered imperative which means is showing us all the different flows of the code.